File: apps/daemon/src/orchestration/automation.ts
/**
 * 人工 DAG 的安全自动推进器。
 *
 * v1 采用整张 Run 共用一个工作区、一次只运行一个 worker。这样下游天然看到
 * 上游写入，又不会把多个写代码 agent 塞进同一目录互相覆盖。等自动 merge 和
 * 冲突处理具备后，才能诚实地开放“每任务 worktree + 并行”。
 */
import { existsSync, realpathSync, statSync } from "node:fs";
import path from "node:path";
import type { AgentKind, ApprovalPolicy } from "@prospero/protocol";
import { DispatchService } from "./dispatch.js";
import type { AutomationWorkspace, Run } from "./model.js";
import { OrchestrationError, OrchestrationStore } from "./store.js";
import { createEsaytree, repoRoot } from "./esaytree.js";
import { WorktreeAssetService } from "./worktree-assets.js";

export interface StartAutomationInput {
  runId: string;
  agent: AgentKind;
  accountId?: string;
  approvalPolicy: ApprovalPolicy;
  workspace: AutomationWorkspace;
  cwd: string;
}

export class AutomationError extends Error {
  constructor(
    message: string,
    readonly code: "automation_invalid" | "not_a_repo",
  ) {
    super(message);
    this.name = "AutomationError";
  }
}

export class AutomationService {
  /**
   * 同一 Run 的 tick 串成 promise 链，防止 task.done 与手机操作同时派发两个 worker；
   * key 是 runId，因此不同 Run 没有全局队列，可同时推进独立编排流。
   */
  private readonly queues = new Map<string, Promise<void>>();

  constructor(
    private readonly store: OrchestrationStore,
    private readonly dispatch: DispatchService,
    private readonly worktreeAssets = new WorktreeAssetService(store),
  ) {}

  async start(input: StartAutomationInput): Promise<Run> {
    const run = this.store.getRun(input.runId);
    if (run.coordinatorSessionId !== null) {
      throw new AutomationError("协调者 Run 已由协调 agent 调度，不能再启用静态自动执行", "automation_invalid");
    }
    if (run.status !== "active") {
      throw new AutomationError(`Run 当前是 ${run.status}，不能启动自动执行`, "automation_invalid");
    }
    if (this.store.listTasks(run.id).length === 0) {
      throw new AutomationError("任务图为空，至少添加一个任务后再运行", "automation_invalid");
    }

    const requestedCwd = path.resolve(input.cwd.trim());
    const cwd = canonicalDirectory(requestedCwd);
    if (!cwd) {
      throw new AutomationError(`项目目录不存在或不是目录: ${requestedCwd}`, "automation_invalid");
    }

    let workspacePath = cwd;
    let branch: string | null = null;
    const existing = run.automation;
    if (input.workspace === "run") {
      // daemon 在建完 worktree、记下 operationId 之前崩溃时，重试要复用已有目录，
      // 不能悄悄再建一份新的集成分支。
      if (
        existing?.workspace === "run" &&
        existing.cwd === cwd &&
        isDirectory(existing.workspacePath)
      ) {
        workspacePath = existing.workspacePath;
        branch = existing.branch;
      } else {
        const repo = await repoRoot(cwd);
        if (!repo) {
          throw new AutomationError(`${cwd} 不在 git 仓库中，不能创建 Run worktree`, "not_a_repo");
        }
        const stamp = Date.now().toString(36);
        const name = `auto-${run.id}-${stamp}`;
        branch = `prospero/${run.id}/auto-${stamp}`;
        const created = await createEsaytree({ repo, name, branch });
        // 创建成功即登记；随后 setRunAutomation 或派发失败时仍能从资产清单找到它。
        this.worktreeAssets.registerRun({
          runId: run.id,
          repo,
          path: created.path,
          branch: created.branch,
        });
        // 用户可能选的是 monorepo 子目录；新 worktree 仍应从对应子目录启动 agent。
        workspacePath = path.join(created.path, path.relative(repo, cwd));
      }
    }

    const now = Date.now();
    this.store.setRunAutomation(run.id, {
      state: "running",
      agent: input.agent,
      ...(input.accountId ? { accountId: input.accountId } : {}),
      approvalPolicy: input.approvalPolicy,
      workspace: input.workspace,
      cwd,
      workspacePath,
      branch,
      startedAt: existing?.startedAt ?? now,
      updatedAt: now,
      lastError: null,
    });
    await this.wake(run.id);
    return this.store.getRun(run.id);
  }

  pause(runId: string): Run {
    const run = this.store.getRun(runId);
    const automation = run.automation;
    if (!automation) {
      throw new AutomationError("这个 Run 尚未启用自动执行", "automation_invalid");
    }
    if (automation.state === "completed") return run;
    return this.store.setRunAutomation(runId, {
      ...automation,
      state: "paused",
      updatedAt: Date.now(),
      lastError: null,
    });
  }

  /** worker 交付、图变更或 daemon 恢复后触发；调用方无需等待下一 worker 建完。 */
  kick(runId: string): void {
    void this.wake(runId).catch((error) => {
      console.warn(
        `[prosperod] 自动编排 ${runId} 推进失败: ${error instanceof Error ? error.message : String(error)}`,
      );
    });
  }

  /** 测试与显式控制路径可等待本轮推进完成。 */
  async advance(runId: string): Promise<void> {
    await this.wake(runId);
  }

  /** daemon 启动后续跑之前处于 running 的静态 Run。 */
  resumePersisted(): void {
    for (const run of this.store.listRuns()) {
      if (run.automation?.state === "running") this.kick(run.id);
    }
  }

  private async wake(runId: string): Promise<void> {
    const previous = this.queues.get(runId) ?? Promise.resolve();
    const next = previous.catch(() => {}).then(() => this.tick(runId));
    this.queues.set(runId, next);
    try {
      await next;
    } finally {
      if (this.queues.get(runId) === next) this.queues.delete(runId);
    }
  }

  private async tick(runId: string): Promise<void> {
    const run = this.store.getRun(runId);
    const automation = run.automation;
    if (!automation || automation.state !== "running" || run.status !== "active") return;

    const tasks = this.store.listTasks(runId);
    if (tasks.length === 0) {
      this.pauseWithError(runId, "任务图为空，自动执行已暂停");
      return;
    }

    const active = this.store.listDispatches(runId).some(
      (candidate) => candidate.state === "starting" || candidate.state === "running",
    );
    if (active) return;

    if (tasks.every((task) => task.status === "done")) {
      try {
        // 只能经由 Run 的唯一完成入口收口。它会再次验证 Gate、Dispatch 与
        // automation 状态，并在成功时原子地将 automation 标成 completed。
        this.store.completeRun(runId, { fromAutomation: true });
      } catch (error) {
        // run-level Gate 不会改变 task 状态；它被解决时会 kick 本 Run，届时
        // 再尝试完成即可。这里不是自动执行故障，也不能把它提前标成 completed。
        if (error instanceof OrchestrationError && error.code === "run_not_completable") return;
        throw error;
      }
      return;
    }

    const stopped = tasks.find(
      (task) => task.status === "failed" || task.status === "blocked" || task.status === "cancelled",
    );
    if (stopped) {
      this.pauseWithError(runId, `任务“${stopped.title}”处于 ${stopped.status}，请处理后继续`);
      return;
    }

    const next = this.store.listReadyTasks(runId)[0];
    if (!next) {
      this.pauseWithError(runId, "当前没有可运行任务；请检查依赖或任务状态");
      return;
    }

    try {
      await this.dispatch.startWorker({
        taskId: next.id,
        agent: automation.agent,
        ...(automation.accountId ? { accountId: automation.accountId } : {}),
        worktree: "none",
        cwd: automation.workspacePath,
        approvalPolicy: automation.approvalPolicy,
      });
    } catch (error) {
      this.pauseWithError(
        runId,
        `派发“${next.title}”失败: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }

  private pauseWithError(runId: string, message: string): void {
    const run = this.store.getRun(runId);
    const automation = run.automation;
    if (!automation) return;
    this.store.setRunAutomation(runId, {
      ...automation,
      state: "paused",
      updatedAt: Date.now(),
      lastError: message,
    });
  }
}

function isDirectory(value: string): boolean {
  if (!existsSync(value)) return false;
  try {
    return statSync(value).isDirectory();
  } catch {
    return false;
  }
}

function canonicalDirectory(value: string): string | null {
  if (!isDirectory(value)) return null;
  try {
    return realpathSync.native(value);
  } catch {
    return null;
  }
}

File: apps/daemon-rs/src/orchestration/automation.rs
//! Static DAG automation runner: one run-wide workspace, one worker at a time.

use std::path::{Path, PathBuf};

use crate::database::{now, validate_text};
use crate::error::{Error, Result};
use crate::worker::Database;

use super::gitops::{create_worktree, repo_root, worktree_default_path};
use super::workers::start_worker;
use super::{
    AutomationState, AutomationWorkspace, RegisterWorktree, Run, RunAutomation, RunStatus,
    StartAutomation, StartWorker, TaskStatus, WorktreeAssetKind,
};
use crate::agent::Agents;

fn canonical_directory(path: &str) -> Result<String> {
    let requested = PathBuf::from(path.trim());
    if !requested.is_absolute() {
        return Err(Error::Invalid("cwd must be absolute".into()));
    }
    let meta = std::fs::metadata(&requested).map_err(|_| {
        Error::Invalid(format!(
            "project directory does not exist: {}",
            requested.display()
        ))
    })?;
    if !meta.is_dir() {
        return Err(Error::Invalid(format!(
            "project path is not a directory: {}",
            requested.display()
        )));
    }
    Ok(std::fs::canonicalize(&requested)
        .unwrap_or(requested)
        .to_string_lossy()
        .into_owned())
}

fn relative_to(repo: &Path, cwd: &Path) -> PathBuf {
    cwd.strip_prefix(repo)
        .unwrap_or(Path::new(""))
        .to_path_buf()
}

pub async fn start_automation(
    database: &Database,
    agents: &Agents,
    input: StartAutomation,
) -> Result<Run> {
    validate_text(&input.approval_policy, 64, false)?;
    if !matches!(
        input.approval_policy.as_str(),
        "strict" | "standard" | "yolo"
    ) {
        return Err(Error::Invalid("approvalPolicy is invalid".into()));
    }
    if !matches!(
        input.agent,
        crate::protocol::AgentKind::Claude | crate::protocol::AgentKind::Codex
    ) {
        return Err(Error::Invalid(
            "automation currently supports Claude/Codex workers".into(),
        ));
    }
    let cwd = canonical_directory(&input.cwd)?;
    let (run, tasks) = database
        .call({
            let run_id = input.run_id.clone();
            move |store| {
                let run = store.orch_run(&run_id)?;
                let tasks = store.list_tasks(Some(&run_id))?;
                Ok((run, tasks))
            }
        })
        .await?;
    if run.coordinator_session_id.is_some() {
        return Err(Error::Invalid(
            "coordinator runs cannot enable static automation".into(),
        ));
    }
    if run.status != RunStatus::Active {
        return Err(Error::Invalid("run is not active".into()));
    }
    if tasks.is_empty() {
        return Err(Error::Invalid("task graph is empty".into()));
    }

    let mut workspace_path = cwd.clone();
    let mut branch = None;
    if input.workspace == AutomationWorkspace::Run {
        if let Some(existing) = &run.automation
            && existing.workspace == AutomationWorkspace::Run
            && existing.cwd == cwd
            && Path::new(&existing.workspace_path).is_dir()
        {
            workspace_path = existing.workspace_path.clone();
            branch = existing.branch.clone();
        }
        if branch.is_none() {
            let repo = repo_root(Path::new(&cwd))?
                .ok_or_else(|| Error::Invalid(format!("{cwd} is not in a git repository")))?;
            let stamp = crate::database::now().to_string();
            let name = format!("auto-{}-{stamp}", run.id);
            let branch_name = format!("prospero/{}/auto-{stamp}", run.id);
            let path = worktree_default_path(&repo, &name);
            create_worktree(
                &repo,
                &super::WorktreeCreate {
                    path: path.clone(),
                    branch: branch_name.clone(),
                },
            )?;
            let created_path = path.to_string_lossy().into_owned();
            database
                .call({
                    let run_id = run.id.clone();
                    let repo_s = repo.to_string_lossy().into_owned();
                    let path = created_path.clone();
                    let branch = Some(branch_name.clone());
                    move |store| {
                        store
                            .register_worktree_asset(RegisterWorktree {
                                kind: WorktreeAssetKind::Run,
                                run_id,
                                task_id: None,
                                repo: repo_s,
                                path,
                                branch,
                            })
                            .map(|_| ())
                    }
                })
                .await?;
            let subdir = relative_to(&repo, Path::new(&cwd));
            workspace_path = Path::new(&created_path)
                .join(subdir)
                .to_string_lossy()
                .into_owned();
            branch = Some(branch_name);
        }
    }

    let timestamp = now();
    database
        .call({
            let run_id = input.run_id.clone();
            let automation = RunAutomation {
                state: AutomationState::Running,
                agent: input.agent,
                account_id: input.account_id,
                approval_policy: input.approval_policy,
                workspace: input.workspace,
                cwd,
                workspace_path,
                branch,
                started_at: run
                    .automation
                    .as_ref()
                    .map(|value| value.started_at)
                    .unwrap_or(timestamp),
                updated_at: timestamp,
                last_error: None,
            };
            move |store| {
                store
                    .set_run_automation(&run_id, Some(automation))
                    .map(|_| ())
            }
        })
        .await?;
    tick_automation(database, agents, &input.run_id).await?;
    let run_id = input.run_id;
    database.call(move |store| store.orch_run(&run_id)).await
}

pub async fn pause_automation(database: &Database, run_id: &str) -> Result<Run> {
    let run_id = run_id.to_owned();
    database
        .call(move |store| store.pause_automation_with_error(&run_id, None))
        .await
}

pub async fn kick_automation(database: &Database, agents: &Agents, run_id: &str) {
    if let Err(error) = tick_automation(database, agents, run_id).await {
        let run_id = run_id.to_owned();
        let _ = database
            .call(move |store| {
                store
                    .pause_automation_with_error(
                        &run_id,
                        Some(format!("automation tick failed: {error}")),
                    )
                    .map(|_| ())
            })
            .await;
    }
}

pub async fn tick_automation(database: &Database, agents: &Agents, run_id: &str) -> Result<()> {
    let run_id = run_id.to_owned();
    let (run, tasks, ready, dispatches) = database
        .call({
            let run_id = run_id.clone();
            move |store| {
                Ok((
                    store.orch_run(&run_id)?,
                    store.list_tasks(Some(&run_id))?,
                    store.list_ready_tasks(&run_id)?,
                    store.list_dispatches(Some(&run_id))?,
                ))
            }
        })
        .await?;
    let Some(automation) = run.automation.clone() else {
        return Ok(());
    };
    if automation.state != AutomationState::Running || run.status != RunStatus::Active {
        return Ok(());
    }
    if tasks.is_empty() {
        pause_with_error(database, &run_id, "任务图为空，自动执行已暂停").await?;
        return Ok(());
    }
    if dispatches.iter().any(|dispatch| dispatch.state.active()) {
        return Ok(());
    }
    if tasks.iter().all(|task| task.status == TaskStatus::Done) {
        database
            .call(move |store| store.complete_run_from_automation(&run_id).map(|_| ()))
            .await?;
        return Ok(());
    }
    if let Some(stopped) = tasks.iter().find(|task| {
        matches!(
            task.status,
            TaskStatus::Failed | TaskStatus::Blocked | TaskStatus::Cancelled
        )
    }) {
        pause_with_error(
            database,
            &run_id,
            &format!(
                "任务“{}”处于 {}，请处理后继续",
                stopped.title,
                stopped.status.label()
            ),
        )
        .await?;
        return Ok(());
    }
    let Some(next) = ready.first() else {
        pause_with_error(
            database,
            &run_id,
            "当前没有可运行任务；请检查依赖或任务状态",
        )
        .await?;
        return Ok(());
    };
    let input = StartWorker {
        task_id: next.id.clone(),
        agent: automation.agent,
        cwd: automation.workspace_path.clone(),
        worktree: "none".into(),
        approval_policy: Some(automation.approval_policy.clone()),
        account_id: automation.account_id.clone(),
        operation_id: Some(format!("automation-{run_id}-{}", next.id)),
    };
    if let Err(error) = start_worker(database, agents, input).await {
        pause_with_error(
            database,
            &run_id,
            &format!("派发“{}”失败: {error}", next.title),
        )
        .await?;
    }
    Ok(())
}

async fn pause_with_error(database: &Database, run_id: &str, message: &str) -> Result<()> {
    let run_id = run_id.to_owned();
    let message = message.to_owned();
    database
        .call(move |store| {
            store
                .pause_automation_with_error(&run_id, Some(message))
                .map(|_| ())
        })
        .await
}
